File: 1345-jump-game-iv/1345-jump-game-iv.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        int N = arr.size();
        unordered_map<int, vector<int>> graph;
        unordered_set<int> visited;
        queue<int> Q;
        int level = 0;
        
        visited.insert(0);
        Q.push(0);
        
        for(int i = 0; i < N; i++){
            graph[arr[i]].push_back(i);
        }
        
        while(!Q.empty()){
            for(int size = Q.size(); size > 0; size--){
                int curr = Q.front(); Q.pop();
                
                if(curr == N - 1)
                    return level;
                
                vector<int>& next = graph[arr[curr]];
                next.push_back(curr + 1);
                next.push_back(curr - 1);
                
                for(int neigh : next){
                    if(0 <= neigh && neigh < N && visited.find(neigh) == visited.end() ){
                        visited.insert(neigh);
                        Q.push(neigh);
                    }
                }
                next.clear();
            }
            
            level++;
        }
        
        return level;
        
        
    }
};

File: 1345-jump-game-iv/1345-jump-game-iv.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        
        N = len(arr)
        graph = defaultdict(list)
        visited = set([0])
        queue = deque([0])      
        level = 0
        
        
        for i in range(N):
            graph[arr[i]].append(i)
        
        
        while queue:
            size = len(queue)
            for _ in range(size):
                
                curr = queue.popleft()
                if curr == N - 1:
                    return level
                
                neighbours = [curr - 1, curr + 1]
                neighbours.extend(graph[arr[curr]])
                for neighbour in neighbours:
                    if 0 <= neighbour < N and neighbour not in visited:
                        visited.add(neighbour)
                        queue.append(neighbour)
                graph[arr[curr]] = []
            level += 1
        
        